### DISCUSSION TITLE
ImageGlass Moon 9.0 beta 5 🌕

### DISCUSSION THREAD

**Discussion by d2phap** [d2phap/ImageGlass#1618] [General]
<img width="640" alt="ImageGlass 9.0 beta 5" src="https://github.com/ImageGlass/website-content/blob/main/releases/screenshots/v9.0-beta-5/9.0b5_1.jpg?raw=true">

[![Downloads](https://img.shields.io/github/downloads/d2phap/imageglass/9.0.5.726/total?color=%232A7C91&label=downloads&style=for-the-badge)](https://imageglass.org/release/imageglass-9.0-beta-5-43)

## ⚠ Notes
This is the final beta release of ImageGlass version 9. Features and settings may still be changed when the official version 9 is released.

To follow the progress of version 9 RC, please visit: https://github.com/d2phap/ImageGlass/projects/16.
You also can support this project by making a donation or purchasing the Microsoft Store version (version 8):

<a href="https://www.microsoft.com/store/productId/9N33VZK3C7TH?cid=GitHubRelease&referrer=appbadge&source=GithubD2phap">
  <img height="69" src="https://user-images.githubusercontent.com/3154213/197671083-4099d80f-07fc-4af9-b80d-6357f0968347.svg" />
</a>

## 🎐 Minimum requirements
- 🆕 Windows 10 version 2004 (20H1) - 64-bit
- 🆕 [.NET Desktop Runtime 6](https://dotnet.microsoft.com/en-us/download/dotnet/6.0)
- 🆕 [WebView2 Runtime x64](https://developer.microsoft.com/en-us/microsoft-edge/webview2/#download-section)

# 📝 Changelog:
ImageGlass 9.0 beta 5 is the continuation of the version 9.0 beta 4,
you should read the previous changelog here: https://github.com/d2phap/ImageGlass/releases/tag/9.0.4.416.

<details>
<summary>

### Show / hide details

</summary>

## 🟣 Breaking changes
- Change app settings and app registry from per-machine to **per-user**:
  + App settings: Use `%LocalAppData%` folder (previouly, it was `%AppData%`) to save user setting file `igconfig.json` when ImageGlass does not have write permission in its startup folder
  + App registry: Use `HKEY_CURRENT_USER` (previously, it was `HKEY_LOCAL_MACHINE`) for registering file type associations, app protocol (#1581)
- Updated the argument naming of the settings `ToolbarItems`, `Tools` from `Argument` to `Arguments`.

## 🟣 General changes
- Upgraded Magick.NET to [version 13.2.0](https://github.com/dlemstra/Magick.NET/releases/tag/13.2.0)
- Removed `igcmd10.exe` utility
- Added support for reading grayscale DNG images (#1575)
- Added support for showing file thumbnail when dragging from Windows to ImageGlass (.NET 7 build on Windows 11)
- Added new commands for `igcmd.exe` utility: `install-languages`, `install-themes`, `uninstall-theme`
- Improved DPI and text scaling for menu
- Improved the direction of the menu dropdown when it's open
- Improved keyboard parser with better accuracy (#1612)
- Improved the stability of Share dialog
- Added setting `SlideshowImagesToNotifySound` to let user determine the number of images to play notification sound in slideshow mode
- Fixed an issue where user could not start slideshow due to incorrect command line prefix
- Fixed an issue where the image is sometimes not displayed after a very quick navigation
- Fixed an issue where orientation of HEIC format is displayed wrong (#1551)
- Fixed an issue where NEF files appear darker with colors less saturated (#1577)
- Fixed an issue where ImageGlass displays NEF files in wrong orientation (#1478)
- Fixed an issue where ImageGlass could not display TIF files due to invalid tag 32932 (#1583) and tag 34031 (#1617)
- Fixed an issue when triggering custom hotkeys such as Alt and key causes windows error sound to play (#1613)

## 🟣 Add UI for App Settings
<img width="400" align="right" src="https://github.com/d2phap/ImageGlass/assets/3154213/5e1268c4-a1dc-4e56-81f0-807c5cccd657" />

- Added support for changing user settings for:
  + Tab General
  + Tab Image
  + Tab Slideshow
  + Tab Edit: except for the `EditApps` setting
  + Tab Viewer
  + Tab Toolbar: except for the `ToolbarItems` setting
  + Tab Gallery
  + Tab Layout
  + Tab Mouse: except for the `MouseClickActions` setting
  + Tab Tools
  + Tab Language
  + Tab Appearance
- Resolved an issue where settings are hidden on high DPI screen (#793, #1266)
- Resolved an issue where emoji characters are not rendered in Settings window (#1220)
- Resolved an issue where user does not see the changes when pressing "Apply" button (#623)
- Resolved an issue where the Settings UI is reset when switching tabs (#482)

## 🟣 Viewer component
- Added support for touch guestures:
  + Supports zooming image by 2 fingers (#591, #739, #1611)
  + Supports panning image with inertia by 1 finger (#591)
  + Supports opening context menu by long tap
- Added support for custom zoom levels via setting `ZoomLevels` (#1563).

### 🟢 Added a new Webview2 viewer component using [HapplaJs library](https://github.com/d2phap/happlajs)
<img height="350" align="right" src="https://github.com/d2phap/ImageGlass/assets/3154213/d7f24ae8-5b77-4c50-8183-48928de78299" />

- Supports scaling SVG (#421), animated SVG (#1098), interactive SVG files
- Supports opening image by dropping files from desktop
- Supports navigation arrow buttons according to the setting
- Supports showing in-app message
- Supports customizing MouseWheel action
- Supports customizing MouseClick action (right button)
- Supports zooming features:
  + All zoom modes
  + Zoom by hotkeys
  + Set custom zoom
  + Follow zoom levels setting
- Supports panning features:
  + Pan by hotkeys
  + Follow pan speed setting
- Other features ***may*** just work, except for Rotation, Flipping, Color picker, Cropping, Page navigation, Copy image data, Print, Set as Desktop/Lock screen, View image channels, Export image frames, ...
- Added an option to switch back to the native viewer component:
  + Open Settings > Viewer
  + Uncheck the option `Use Webview2 for viewing SVG format` (or set `UseWebview2ForSvg` to `false` in `igconfig.json`).
- When ImageGlass window is minimized, the Webview2 viewer component will be suspended to consume less memory similar to MS Edge's sleep mode.

## 🟣 Toolbar component
- Added support for adding external tool as toolbar button using the tool ID
- Improved Main menu position to make sure it does not cover the toolbar when the toolbar is at bottom
- Fixed an issue where user cannot use custom image path for toolbar icon (#1554)
- Fixed an issue where user cannot click the toolbar buttons when the window is inactive (#1202, #1363)
- Fixed an issue where tooltip overlaps the hovered button when the gallery's position is not on top

## 🟣 Gallery component:
- Added support scrolling thumbnails by swipe guesture (#753)
- Added support for resizing gallery by dragging the separator in vertical layout
- Added support for saving the last size of gallery component (#1502)
- Added support for displaying scrollbars with dark mode
- Improved gallery layout scaling according to system text size setting
- Improved click action to only allow to use the left mouse for image navigation
- Fixed an issue where the dragged thumbnail is still selected after mouse is released
- Fixed an issue where thumbnail's tooltip is still showing after leaving the thumbnail
- Fixed an issue where tooltip overlaps the hovered thumbnail when the gallery's position is not at bottom

## 🟣 Tooltip component
<img height="150" align="right" src="https://github.com/d2phap/ImageGlass/assets/3154213/f8a9e40c-4221-4e65-930e-bc0a798dcc19" />

- Added supports for dark mode
- Added supports round corner on Windows 11
- Added supports for using accent color for tooltip title
- Improved tooltip size and text scaling according to system text size setting

## 🟣 ImageGlass Tools
- Added support for CIELAB color space for Color Picker tool (#1543)
- Added support for setting topmost state of tool window according to ImageGlass window
- Added support for assigning hotkeys for external tool in `Tools` section in `igconfig.json` file:
  ```js
  // in igconfig.json
  "Tools": [
    {
      "ToolId": "Tool_MyDemoApp", // a unique ID
      "ToolName": "My Demo app", // name of the tool
      "Executable": "path\\to\\the\\DemoApp.exe",
      "Arguments": "<file>", // file path to pass to the tool
      "Hotkeys": ["X", "Ctrl+E"], // press X or Ctrl+E to open/close the tool
      "IsIntegrated": true|false // true: if the tool supports 'ImageGlass.Tools'
    }
  ]
  ```
- Improved the built-in tool window readability by increasing opacity to 95% (from 85%) in inactive state
- Fixed an issue where the second ImageGlass instance cannot open the external tool if other instances already open it

</details>

## 🟣 Other updates for ImageGlass 9.0 beta 5
- Published [ImageGlass.Tools library v1.9050](https://github.com/ImageGlass/ImageGlass.Tools)
- Published [ExifGlass app v1.1](https://github.com/d2phap/ExifGlass)

## 🥒 Checksums:

File | SHA-1
-- | --
ImageGlass_Moon_9.0.5.726_Beta-5_x64_net6.zip | `CEDD125A9581B91D487018C71331D1C867A43B56`
ImageGlass_Moon_9.0.5.726_Beta-5_x64_net7.zip | `1545D69391F99686E0189CAC0549B891C1D14ED5`

**Full Changelog**: https://github.com/d2phap/ImageGlass/compare/9.0.4.416...9.0.5.726

<hr /><em>This discussion was created from the release <a href='https://github.com/d2phap/ImageGlass/releases/tag/9.0.5.726'>ImageGlass Moon 9.0 beta 5 🌕</a>.</em>

**Comment by ghost** [comment]
I check this update and its very nice.......

are you adding this in setting panal?
Your Comment Refrence: https://github.com/d2phap/ImageGlass/issues/971#issuecomment-1133560019

**Reply by d2phap** [reply]
I totally forgot about this setting! I will add in the RC release.

**Comment by ghost** [comment]
Double Click Mouse Option Not Worked or Not Implement 
.
.

In Old Version: v8x v7x .......

If Zoom Out Image
Double Click = 100% Zoom

If Zoom In or Zoom Grater then ImageGlass Window Hight/Wight
Double Click = Fit To ImageGlass Window W/H

**Reply by d2phap** [reply]
the default action for double click is just not implemented yet

**Comment by Radnarsen** [comment]
It'll be convenient if the frameless mode leaves an active one after returning from the full-screen (F11).

**Reply by d2phap** [reply]
thanks for pointing out, it will be fixed in the next release.

**Comment by mojoZHU** [comment]
The settings UI panel is not functional for me. I have tried the .NET 6 and .NET 7 versions. I can confirm that the ipconfig.json file is being created in the root directory of the program
![Screenshot 2023-08-08 162311](https://github.com/d2phap/ImageGlass/assets/83429154/20b7b94a-cb32-4736-8b93-68a60a2b103a)

**Reply by d2phap** [reply]
There might be JavaScript error in Settings UI. What about other windows such as About dialog, Check for update dialog and the SVG viewer?

**Reply by mojoZHU** [reply]
Those seem to work fine and the buttons work too, unlike the Settings UI. 

The beta is pretty good in general and I use it as my only image viewer. Keep up the good work. I only want to help if I can.
![Screenshot 2023-08-09 104404](https://github.com/d2phap/ImageGlass/assets/83429154/d50c20a9-502d-4300-b53c-06bdd5909511)

**Reply by d2phap** [reply]
Thanks for trying. There is currently no way to check what the error is because I totally disabled Webview2's DevTool for the release build 🥲.

I will create a setting to enable it via `igconfig.json` file.

**Reply by mojoZHU** [reply]
Okay. I think my settings are still being saved. An example is if I switch the program to "frameless" mode, it will open the next time frameless.

**Reply by d2phap** [reply]
Btw, I manage to reproduce the issue. It seems that you extracted ImageGlass to Program File folder which requires admin permission to write. You can try set write permission to the ImageGlass folder, it should work!

**Reply by d2phap** [reply]
The root cause is because the Settings window tries to load all svg icons from `Icons` folder. This folder is located in the startup folder, but it looks for the Config folder. In your case, the Config folder (`%LocalAppData%\ImageGlass`) is not the same as the startup folder.

I've just pushed a fix #1630

**Comment by Radnarsen** [comment]
When I turn off the option "Show gallery scrollbars", the scrollbar shows when using the side gallery arrangement anyway. Is it the bug or must it be like this?

**Reply by d2phap** [reply]
It's by design. Similar to v8, vertical gallery will always show the vertical scrollbar

**Comment by ghost** [comment]
I tested it and didn't notice any particular issues in the UI, but what I really don't like is the image loading time. I was hoping that this version might show certain improvement in that area, but sadly I must note that's not the case. All in all, I was really excited about this application for a while, but the biggest problem that I have with it at the moment is in the fact that Image Glass is actually way too slow during the loading of larger images in resolution in comparison with Windows Image Viewer, or particularly Pictus, which is even faster. Pictus is also very useful because it can activate PSD thumbnail previews inside the Explorer, but what is more important, it can open very large PSD files almost instantly. So my friendly advice to developers would be next. Beside the fact that you made nice looking lightweight image viewer, what you should really try to do now is to make the fastest image viewer, because otherwise users will start to search for alternatives, similarly like I did already. I really like this application, but because of all mentioned, I'm basically forced to search alternative solution. Keep on good work.
